channels/pkg/channels/addon.go

```diff
@@ -161,17 +168,90 @@ func (a *Addon) EnsureUpdated(ctx context.Context, k8sClient kubernetes.Interfac
 	}
 
 	if required.NewVersion != nil {
+
 		manifestURL, err := a.GetManifestFullUrl()
 		if err != nil {
 			return nil, err
 		}
 		klog.Infof("Applying update from %q", manifestURL)
 
-		err = Apply(manifestURL.String())
+		manifestPath, err := vfs.Context.BuildVfsPath(manifestURL.String())
+		if err != nil {
+			return nil, fmt.Errorf("failed to build manifest path: %w", err)
+		}
+
+		targetManifest, err := manifestPath.ReadFile()
 		if err != nil {
-			return nil, fmt.Errorf("error applying update from %q: %v", manifestURL, err)
+			return nil, fmt.Errorf("error reading manifest: %v", err)
 		}
 
+		appliedUrl, _ := path.Split(manifestPath.Path())
+		appliedUrl = appliedUrl + "applied.yaml"
+		appliedPath, err := vfs.Context.BuildVfsPath(appliedUrl)
+		if err != nil {
+			return nil, fmt.Errorf("failed to build applied manifest path: %w", err)
+		}
+
+		target, err := kube.Build(bytes.NewBuffer(targetManifest), true)
+		if err != nil {
+			return nil, err
+		}
+
+		currentManifest, err := appliedPath.ReadFile()
+		if err != nil {
+			if os.IsNotExist(err) {
+				klog.Infof("installing new addon %q", required.Name)
+				currentManifest = []byte{}
+			} else {
+				return nil, err
+			}
+		}
+
+		current, err := kube.Build(bytes.NewBuffer(currentManifest), true)
+		if err != nil {
+			return nil, err
+		}
+
+		existingResources := make(map[string]bool)
+		for _, r := range current {
+			existingResources[objectKey(r)] = true
+		}
+
+		var toBeUpdated helmkube.ResourceList
+		{
+			err := target.Visit(func(info *resource.Info, err error) error {
+				{
+					helper := resource.NewHelper(info.Client, info.Mapping)
+					_, err := helper.Get(info.Namespace, info.Name)
+					if err != nil {
+						if errors.IsNotFound(err) {
+							// This is the happy path. Resource does not exist already.
+							return nil
+						}
+						return err
+					}
+
+					toBeUpdated.Append(info)
+					return nil
+				}
+			})
+
+			if err != nil {
+				return nil, fmt.Errorf("unable to get information about existing resources: %w ", err)
+			}
+		}
+
+		for _, r := range toBeUpdated {
+			current.Append(r)
```

Comment:
I'm not sure I follow what's going on here ... is current the version on the cluster or in the previously applied yaml?  Is helm merging the two by gvknn?

Reply:
The previously applied yaml. This comes from the challenge I mentioned with being able to extract existing resources from k8s directly. So both I and Helm rely on the state in k8s being the same as the state from the previously applied yaml.